jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2StreamEndPoint.java

```diff
@@ -248,92 +249,27 @@ public boolean flush(ByteBuffer... buffers) throws IOException
-                        // write(), the semantic of flush() is that it does not
-                        // own them, but stream.data() needs to own them.
-                        ByteBuffer buffer = coalesce(buffers, true);
-                        Callback.Completable callback = new Callback.Completable(Invocable.InvocationType.NON_BLOCKING);
-                        stream.data(new DataFrame(stream.getId(), buffer, false), callback);
-                        callback.whenComplete((nothing, failure) ->
-                        {
-                            if (failure == null)
-                                flushSuccess();
-                            else
-                                flushFailure(failure);
-                        });
-                        return callback.isDone();
-                    case PENDING:
-                        return false;
-                    case OSHUTTING:
-                    case OSHUT:
-                        throw new EofException("Output shutdown");
-                    case FAILED:
-                        Throwable failure = current.failure;
-                        if (failure instanceof IOException)
-                            throw (IOException)failure;
-                        throw new IOException(failure);
-                }
-            }
-        }
-    }
-
-    private void flushSuccess()
-    {
-        while (true)
-        {
-            WriteState current = writeState.get();
-            switch (current.state)
-            {
-                case IDLE:
-                case OSHUT:
-                    throw new IllegalStateException();
-                case PENDING:
-                    if (!writeState.compareAndSet(current, WriteState.IDLE))
-                        break;
-                    return;
-                case OSHUTTING:
-                    shutdownOutput();
-                    return;
-                case FAILED:
-                    return;
-            }
-        }
-    }
 
-    private void flushFailure(Throwable failure)
-    {
-        while (true)
+        WriteState current = writeState.get();
+        switch (current.state)
         {
-            WriteState current = writeState.get();
-            switch (current.state)
-            {
-                case IDLE:
-                case OSHUT:
-                    throw new IllegalStateException();
-                case PENDING:
-                    if (!writeState.compareAndSet(current, new WriteState(WriteState.State.FAILED, failure)))
-                        break;
-                    return;
-                case OSHUTTING:
-                    shutdownOutput();
-                    return;
-                case FAILED:
-                    return;
-            }
+            // The flush() semantic is that it must not leave pending operations, therefore we
+            // cannot call stream.data() because it would remain pending if the stream is congested.
+            case IDLE:
```

Comment:
shouldn't IDLE be return true?   All data has been flushed because there is no data.

Reply:
@gregw this is the state before doing the "flush", so even when IDLE we do not perform the flush and so we return `false`.

Reply:
Sorry but I totally do not understand this method now?   It is essentially a noop as it neither changes state, nor writes any data???

Reply:
Exactly. It only perform actions when oshut or already failed.